dulwich/porcelain.py

```diff
@@ -1243,6 +1268,15 @@ def prune_dirnames(dirpath, dirnames):
                         os.path.join(os.path.relpath(path, frompath), "")
                     )
                 del dirnames[i]
+            elif not walk:
```

Comment:
`walk` seems like a very generic term that doesn't really describe what is going on. Not that `normal` vs `all` are very descriptive themselves, but maybe let's just pass the policy down here and decide what to do in one place instead of introducing another translation layer with `walk`?